rfcs/20241205-graph-api-causal-mask/README.md

```diff
@@ -0,0 +1,485 @@
+
+![causal mask subgraph top left](images/causal_mask_subgraph_top_left.png)
+
+This structure effectively implements the causal mask, ensuring that each position
+can only attend to its previous positions and itself in the sequence.
+
+#### Option 1.2: Bottom-Right matrix generation
+
+For bottom-right alignment matrix, e.g.
+
+```cpp
+# s_q=5, s_kv=5             |    # s_q=2, s_kv=5             |    # s_q=5, s_kv=2
+  0  -inf -inf -inf -inf    |      0    0    0    0  -inf    |      -inf  -inf
+  0    0  -inf -inf -inf    |      0    0    0    0    0     |      -inf  -inf
+  0    0    0  -inf -inf    |                                |      -inf  -inf
+  0    0    0    0  -inf    |                                |        0   -inf
+  0    0    0    0    0     |                                |        0    0
+```
+
+User can construct the subgraph using the following structure:
+
+1. The `GenIndex` operation generates indexes for rows and columns as before.
+2. The condition for masking is changed (indicated by the yellow blocks in the
+following image), if the column index is less than the row index + s_kv - s_q, it
+indicates future tokens. This condition can be represented with `Greater Than`,
+`Add`, `Sub`and two`Constant` operations.
+3. The `Select` operation sets these future token positions to -inf.
+
+![causal mask subgraph bottom_right](images/causal_mask_subgraph_bottom_right.png)
+
+With this proposal, the SDPA pattern will support above subgraph-based masks:
+
+![sdpa with causal mask subgraph](images/sdpa_w_causal_mask_subgraph.png)
+
+To support this proposal, oneDNN Graph will need to support `GenInex` and
+`GreaterThan` and `Constant` operations additionally.
+
+#### GenIndex Op Support
+
+`GenInex` operation generates an index tensor along a specified axis of a given
+tensor. This operation is useful for creating index-based masks or for other purposes
+where the position of elements within a specific dimension is needed.
+The generated index tensor has the same shape as the input tensor, with values
+corresponding to the indices along the specified axis.
+
+
+| Attribute Name | Description | Value Type | Supported Values | Required or Optional |
+| :------------- | :---------- | :--------- | :--------------- | :------------------- |
+| [axis] (@ref dnnl::graph::op::attr::axis) | Specifies the dimension along which index values are generated. | s64 | A s64 value in the range of [-r, r-1] where r = rank(src) | Required |
+
+##### Execution arguments
+
+##### Inputs
+
+| Index | Argument Name | Required or Optional |
+| :---- | :------------ | :------------------- |
+| 0     | `src`         | Required             |
+
+##### Outputs
+
+| Index | Argument Name | Required or Optional |
+| :---- | :------------ | :------------------- |
+| 0     | `dst`         | Required             |
+
+##### Supported data types
+
+`GenInex` operation supports the following data type combinations.
+
+| src  | dst  |
+| :--- | :--- |
+| f32  | f32  |
```

Comment:
maybe we can add s32/s64 for output?  pytorch index operations only accept s32/s64. such as https://pytorch.org/docs/stable/generated/torch.Tensor.index_add_.html

Reply:
good catch, will add :D

Reply:
But in fact, oneDNN does not support s64 as memory/tensor data type.

Reply:
I also checked the document of CUDNN_POINTWISE_GEN_INDEX, although there's no explicit information about the output data type of this operation, as CUDNN supports INT32 and INT64  data type, we can infer that the op is likely to generate INT32 or INT64, as these are common for representing indices.

Given that, there will be two options for the output data type of GenIndex:

1. Keep f32 data type: if users need integer value, they can convert.
2. Use int32/int64 data type: oneDNN needs to add a new tensor data type, which seems to be overly complex if only consider the specific use case of implicit causal mask.